File: leetcode/374. Guess Number Higher or Lower.cc
// Copyright 2016 Qi Wang
// Date: 2016-11-16
// The declaration is necessary here
int guess(int num);

class Solution {
 public:
  int guessNumber(int n) {
    size_t l = 1, r = n;
    while (l <= r) {
      int mid = (l + r) >> 1;
      int ans = guess(mid);
      if (ans == 0) {
        return mid;
      } else if (ans == 1) {
        l = mid + 1;
      } else {
        r = mid - 1;
      }
    }
    return -1;  // comfort the compiler
  }
};

File: leetcode/374. Guess Number Higher or Lower.java
// Date: 2020-08-02
// loop invariant: answer exists in [left, right]
public class Solution extends GuessGame {
  public int guessNumber(int n) {
    int left = 1;
    int right = n;
    while (left < right) {
      int mid = (right - left) / 2 + left;
      switch (guess(mid)) {
        case 0:
          return mid;
        case -1:
          right = mid - 1;
          break;
        case 1:
          left = mid + 1;
      }
    }
    return left;
  }
}

// Date: 2020-08-02
// loop invariant: answer exists in [left, right]
public class Solution extends GuessGame {
  public int guessNumber(int n) {
    int left = 1;
    int right = n;
    while (left <= right) {
      int mid = (right - left) / 2 + left;
      switch (guess(mid)) {
        case 0:
          return mid;
        case -1:
          right = mid - 1;
          break;
        case 1:
          left = mid + 1;
      }
    }
    return 0; // only used to comfort the compiler
  }
}
